### PR TITLE
feat(v2/storage-5): allow download of responses with date range

### PR DESCRIPTION
**Pull request by karrui**
Builds on #3965

## Problem

This PR allows for downloading of responses by entering a date range. Also updates the method of calculating the number of responses to download by retrieving responses count by date on date range selection instead of only during decryption.

Related to #3129

## Solution

**Features**:

- feat: pass in selected date range to storage download
- feat: pass separate `dateRangeResponsesCount` prop via context
- feat: use `dateRangeResponseCount` for rendering download progress

**Improvements**:

- ref(storage): rename `responsesCount` prop passed down from context to `totalResponsesCount`

## Before & After Screenshots
None needed i think...

### CODE REVIEW

File: frontend/src/features/admin-form/responses/AdminSubmissionsService.ts [hunk 1]
```diff
@@ -18,9 +18,13 @@ export const countFormSubmissions = async ({
   dates,
 }: {
   formId: string
-  dates?: SubmissionCountQueryDto
+  dates?: Partial<SubmissionCountQueryDto>
 }): Promise<number> => {
   const queryUrl = `${ADMIN_FORM_ENDPOINT}/${formId}/submissions/count`
+  if (dates && (!dates.startDate || !dates.endDate)) {
+    delete dates.startDate
+    delete dates.endDate
```
Comment (tshuli) [inline frontend/src/features/admin-form/responses/AdminSubmissionsService.ts]: Why is this necessary; what happens if we simply pass in undefined startDate and endDate? 

Also should we allow a case where only the startDate or endDate is specified? Right now both must be specified to pass the test

Comment (karrui) [reply frontend/src/features/admin-form/responses/AdminSubmissionsService.ts]: following current implementation tbh, server will reject if only one date is supplied

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/StorageResponsesProvider.tsx [hunk 2]
```diff
@@ -29,25 +38,31 @@ export const StorageResponsesProvider = ({
...
+    if (!secretKey || !dateRangeResponsesCount) return null
 
     return {
       secretKey,
-      // TODO: Add selector for start and end dates.
-      endDate: undefined,
-      startDate: undefined,
+      responsesCount: dateRangeResponsesCount,
+      startDate: dateRange[0],
+      endDate: dateRange[1],
```
Comment (timotheeg) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/StorageResponsesProvider.tsx:47]: Just to confirm my understanding: if `dateRangeResponsesCount` is greater than 0, it means for sure `dateRange` is not `undefined`, is that correct?

I'm asking because lines 30+31 indicate `dateRange` _could_ be `undefined`, but by the time we reach this code here, `dateRange` is guaranteed to be defined, and therefore we don't need the optional chaining syntax any more, correct?

Extension to this question: Is there a guarantee that `setDateRange()` will never be called with `undefined`, such that we could change line 30+31 to remove the optional chaining?

File: frontend/src/features/admin-form/responses/AdminSubmissionsService.ts [hunk 3]
```diff
@@ -18,9 +18,13 @@ export const countFormSubmissions = async ({
...
+  }
   if (dates) {
```
Comment (timotheeg) [inline frontend/src/features/admin-form/responses/AdminSubmissionsService.ts]: /nit suggestion

```suggestion
  if (dates && dates.startDate && dates.endDate) {
```

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/StorageResponsesProvider.tsx (hunk 2 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/StorageResponsesProvider.tsx:47]: ah yes, brain fart. will remove optional chain.

File: frontend/src/features/admin-form/responses/AdminSubmissionsService.ts (hunk 3 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/AdminSubmissionsService.ts]: done.

**Review by timotheeg — APPROVED** [review]
lgtm 👍
